croncheck.php

```diff
@@ -284,51 +277,56 @@
         require_once($CFG->dirroot.'/mnet/lib.php');
     }
 
-    $checks = \core\check\manager::get_checks('status');
-    $output = '';
-    // Should this check block emit as critical?
-    $critical = false;
-
-    foreach ($checks as $check) {
-        $ref = $check->get_ref();
-        $result = $check->get_result();
-
-        $status = $result->get_status();
-
-        // Summary is treated as html.
-        $summary = $result->get_summary();
-        $summary = html_to_text($summary, 80, false);
-
-        if ($status == \core\check\result::WARNING ||
-            $status == \core\check\result::CRITICAL ||
-            $status == \core\check\result::ERROR) {
-
-            // If we have an error, how should we handle it.
-            if ($status == \core\check\result::ERROR && !$critical) {
-                $mapping = get_config('tool_heartbeat', 'errorcritical');
-                if ($mapping === 'critical') {
-                    $critical = true;
-                } else if ($mapping === 'criticalbusiness') {
-                    // Here we should only set the critical flag between 0900 and 1700 server time.
-                    $time = new DateTime('now', core_date::get_server_timezone_object());
-                    $hour = (int) $time->format('H');
-                    $critical = ($hour >= 9 && $hour < 17);
+    try {
```

Comment:
Note - this try/catch was all that was added (git diff butchered it unfortunately)

Reply:
Can this try catch be lower down so its specific to a check? Lets say we have 4 failing tasks and 1 broken task, I think this will end up with just a broken message rather than 4 failing + 1 broken?

Reply:
It's possible for `\core\check\manager::get_checks` to also throw an exception, which is something I was intending to catch (i.e. anything that can fail should be caught)

My thinking was that ideally none of these _should_ never throw an exception under normal circumstances. Only when for e.g. a bad plugin is installed with a dodgy class does it throw the exception

Maybe my assumption is wrong ?

Reply:
Honestly we can't be too careful here so lets leave it as is thanks

Reply:
I thought about it more and I think you are actually correct, for example if a check sends a curl request and this dies it might throw an exception, but we don't want this masking other issues (they would disppear, and a dev might think they got fixed)

So now I've reworked it:

- All the actual check logic is inside `$processcheckfn` (I moved it into its own function since the nesting was getting a bit too unruly)
- We just use array_map and try/catch using this function
- If multiple are shown, the first line shows a summary (e.g. 2 checks failed) - heading in the direction of #137 

Example output now (with a task I forced to throw an exception):

```
CRITICAL: 3 status checks have reported warnings or errors: 
Cron running
The admin/cron.php script has not been run for 14 mins 3 secs and should run
every 1 min.
Details: https://moodle-311-heartbeat.localhost/report/status/index.php?detail=tool_task_cronrunning

Check "Tasks max fail delay" threw an exception: coding_exception: Coding error detected, it must be fixed by a programmer: test in /var/www/moodle-311-heartbeat/admin/tool/task/classes/check/maxfaildelay.php:58
Stack trace:
#0 /var/www/moodle-311-heartbeat/admin/tool/heartbeat/croncheck.php(295): tool_task\check\maxfaildelay->get_result()
#1 /var/www/moodle-311-heartbeat/admin/tool/heartbeat/croncheck.php(341): {closure}(Object(tool_task\check\maxfaildelay))
#2 [internal function]: {closure}(Object(tool_task\check\maxfaildelay))
#3 /var/www/moodle-311-heartbeat/admin/tool/heartbeat/croncheck.php(347): array_map(Object(Closure), Array)
#4 {main}

Ad hoc task queue
Oldest task is 16 mins 31 secs which is more than 10 mins
Details: https://moodle-311-heartbeat.localhost/report/status/index.php?detail=tool_task_adhocqueue
 (Checked Oct 9 16:09:56)
```